### PR TITLE
Dev setup updates – spring 2018

### PR DESCRIPTION
**Pull request by borekb**
Our dockerized dev setup, originally built in #1227, was not working for me, for example, the newest Xdebug is no longer compatible with PHP 5.x. When fixing this, I've encountered several other issues with our dev / test setup so this PR ends up being a pretty major update.

### Highlights

- Docker images are now [pushed to Docker Hub](https://hub.docker.com/r/versionpress/wordpress/) as a source of truth. Previously, it was necessary to build images locally which could fetch newer dependencies than originally intended and break things.
- Dev and testing stacks have been unified. There is now a single `docker-compose.yml`, using the same images for both scenarios.
- Dockerized environment has been updated so that tests pass again
    - A couple are still failing (Workflow tests) but that might be due to actual bugs; to be resolved separately.
    - Selenium tests have been disabled entirely, see below.
- Gulp is gone, replaced by simpler npm scripts (252c295).
- PHP used for dev site and tests updated from 5.6 to 7.2.
- Git Bash is required on Windows.
- [`Dev-Setup.md`](https://github.com/versionpress/versionpress/blob/baa15a3c55d360fa868a00396701aefd20de85f5/docs/Dev-Setup.md) has been reviewed and updated.

### Noted for future work

- Make Selenium tests work again. I spent some time on them in this PR but they will require more love. Quick thoughts at this point:
    - The actual test code (navigating URLs, clicking on things) still looks fine to me but the setup code is slightly suspicious, e.g., [here](https://github.com/versionpress/versionpress/blob/d322b8bdf1be5d4d01eaeacbe09877aaa4e3d424/plugins/versionpress/tests/End2End/Utils/SeleniumWorker.php#L17-L57) or [here](https://github.com/versionpress/versionpress/blob/d322b8bdf1be5d4d01eaeacbe09877aaa4e3d424/plugins/versionpress/tests/End2End/Utils/SeleniumWorker.php#L89-L103).
    - `facebook/webdriver` seems to be [more popular](https://packagist.org/?query=selenium) than `giorgiosironi/phpunit-selenium` (`Selenium2TestCase`), I've also heard good things about [Codeception](https://codeception.com/). As long-term maintainability of our Selenium tests is an important goal, considering a new framework is something we should do.
- Generally refresh the code in tests. Most of it is alright, some could be improved.
- `test-config.yml` is a legacy concept, for example, a test suite should define tests for both WP-CLI and Selenium workers, or for multiple versions of WordPress. Picking test suites from `phpunit.xml` should be enough.
- Create a docker-compose version that does not contain any host-mounted volumes. That will very likely speed things up dramatically on Windows and macOS, and will enable us to deploy the test suite on CI.
- Yes, CI.
- VersionPress should bump its required PHP version to at least 7.0, see related discussion below: https://github.com/versionpress/versionpress/pull/1329#issuecomment-390119130
- Various other improvements:
    - Print progress of site setup when running tests (the console is "dark" for quite a while).
    - Better configuration of PhpStorm inspections
    - Adopt PHPStan (we already have an issue for it: #1210)

### Full todo list

- [x] Update dev and tests PHP from 5.6 to 7.2
- [x] Generally review Docker images.
- [x] Push images to Docker Hub, script it.
    - Repo is [`versionpress/wordpress`](https://hub.docker.com/r/versionpress/wordpress/)
    - New `npm run build-images` and `npm run push-images` scripts (eed0177, 7fa6847)
- [x] Explore why tests are failing and fix them
    - Most tests were failing due to outdated stack, e.g., file system permissions were not right.
    - There are still some failing tests, see e.g. https://github.com/versionpress/versionpress/pull/1329#issuecomment-391311790, but that will be for separate PRs.
- [x] Explore Swarm / stacks instead of docker-compose
    - We can't use it. An experiment is recorded in 48ff634 but was removed as Docker Swarm mode lack support for one-off jobs which we need for tests. Related: moby/moby#23880.
- [x] Unite dev-env and testing stack (🎉)
    - Done in 06e1965, makes a lot of things simpler.
- [x] Replace `cleanup-docker-stack` with `npm stop` and `npm run stop-and-cleanup`.
    - Done as part of 06e1965 (unit dev and test stack)
- [x] `phpunit.override.xml` is not a great name, it's not overriding anything. Gitignore `phpunit.*.xml` and update docs.
- [x] Get rid of Gulp as it is most likely causing subtle issues, and is generally a legacy technology anyway
    - Done in 252c295.
- [x] 🏃 General Dev-Setup.md updates, @pavelevap to contribute his experience with Docker Toolbox. _Almost done._
- [x] Think about ports, e.g., 3306 might conflict with local database server.
    - The bigger problem was port 80 which is not easily mapped to something else due to how WordPress treats `home` / `siteurl`. So both web servers and DB servers run on their default ports 80 and 3306. See 920e8fb1.
- [x] Revert changes to frontend build script in https://github.com/versionpress/versionpress/commit/0cf52f0eb3922c9418c99799d389bb0dc310fe9d
- [x] ~~Test with `cmd.exe`~~
    - After removing Gulp (252c295), `cmd.exe` was close to working again but it will just be easier to require Git Bash (`curl`, `rm -rf` etc.). It is automated since 9e0904e.
- [x] @pavelevap is experiencing issues with [symlink in `copyVersionPressFiles()`](https://github.com/versionpress/versionpress/pull/1329/commits/815595f0bb5aeb9a0a57e7445faf94723e521863#diff-a1abcc17c8fe775f21cfa9f60041783bR121), maybe related to Docker Toolbox? Explore this.
    - Symlink reverted in ea68740.
- [x] Check PhpStorm files
- [x] Move certain CLI arguments like `--color` from `package.json` to `phpunit.xml`
- [x] Consider switching to digests – [example](https://github.com/docker/compose/blob/651283eef96a6ab7854b3b8ee57bc25467f27f75/tests/fixtures/simple-composefile/digest.yml)
    - Won't happen now.
- [x] Possibly output progress info during site setup - some steps take tens of seconds and the console is completely dark in that case.
    - Postponed.
- [x] Try removing volumes to speed things up
    - Later.

_(This is an issue-less PR.)_

### CODE REVIEW

**Comment by borekb** [comment]
Current status:

- I am able to `npm install` && `npm start` and browse the dockerized dev site, log in to it and activate VersionPress.
- I am able to run all the unit tests successfully: `cd ./plugins/versionpress/tests` && `docker-compose run tests ../vendor/bin/phpunit -c phpunit.xml --testsuite Unit`
- Full test suite can be run without crashing down but some tests are failing:

image

@pavelevap could you please try it on your computer and generally go through the `Dev-Setup.md` document to see if things make sense and are generally working? Feel free to update the instructions as you see fit and push to this PR.

File: frontend/package.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
   },
   "scripts": {
     "start": "node webpack/webpack-dev-server.js --config webpack/webpack.dev.js --env.port 8888",
-    "build": "npm install && tsc --noEmit && webpack -p --config webpack/webpack.prod.js --hide-modules",
+    "build": "npm install && node_modules/.bin/tsc --noEmit && node_modules/.bin/webpack -p --config webpack/webpack.prod.js --hide-modules",
```
Comment (pavelevap) [inline frontend/package.json]: This change does not work for me with `cmd.exe`.

**Comment by leepeterson** [comment]
@borekb about how soon do you think this’ll be approved and merged? exciting stuff!

**Comment by borekb** [comment]
@leepeterson The tests are currently failing and the updated Dockerfiles are not entirely finished yet. @pavelevap is looking at it, I personally won't have much time to return to it until next week.

**Comment by borekb** [comment]
✨ I've significantly updated our devenv / testing stack in https://github.com/versionpress/versionpress/pull/1329/commits/06e1965db7494a9dfb124ac6266e77f2c23b9af8. See the commit's full description for all the details but the TL;DR is that we now use a **unified Docker Compose stack for both devenv and testing purposes** and the original `testenv` image is gone.

The integration tests are still failing but at least there is now a predictable environment where to hunt bugs down.

**Comment by borekb** [comment]
Killed the permission bugs in 87769bf! 🎉 

Many more tests pass now:

image

**Comment by borekb** [comment]
Alright, many improvements done, I won't have much time until late next week to work on this so here is a brief summary of what has happened since the last comment here:

- Tests are much faster to start – we were incorrectly initializing them. 815595f
- There is a concrete fix of VersionPress behaviour around `theme/switch` in fcde513 which, on one hand, is good but on the other hand, I'm not 100% it's correct and maybe it should be reverted and fixed / reviewed in a separate, follow-up PR.
- New `tests:custom` and `tests:custom-with-wordpress` npm scripts, better logging by default and similar "DX" improvements.
- Improvements to Docker images workflow, e.g., better tagging (00c0e73) or scripted push (7fa6847).
- Gulp replaced with simpler npm scripts. 252c295
- Git Bash is strongly recommended on Windows, the `init-dev` script prints a warning if it's not used.
- `Dev-Setup.md` generally reviewed and refreshed – nothing ground-breaking but many smaller updates.
- Cleanup of `.gitignore` files
- Update of PhpStorm config for 2018.1. Unfortunately, debugging tests is currently broken, see [WI-42148](https://youtrack.jetbrains.com/issue/WI-42148) and [IDEA-192128](https://youtrack.jetbrains.com/issue/IDEA-192128).

**Comment by kalich5** [comment]
The wordpress.org repository can not upload plugins with the required minimum version of PHP 7.1

**Comment by borekb** [comment]
@kalich5 Can you provide a source of that information? I don't see any such limitation in [Detailed Plugin Guidelines](https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/) and there also [this make.wordpress.org post](https://make.wordpress.org/plugins/2017/08/29/minimum-php-version-requirement/) about the new `Requires PHP:` field, which, to me, hints that you could do something like `Requires PHP: 7.1` just fine.

**Comment by kalich5** [comment]
https://wordpress.slack.com/files/U0LF07HJP/FAQJJN1C7/40015478-c7d0a5c2-57b3-11e8-8cb3-8b5690b6a9ff.png

**Comment by kalich5** [comment]
The files contain variables from PHP 7.1 which I did not manage to upload to repository on wp.org.

**Comment by borekb** [comment]
Thanks, that's interesting. Here's a [link to the whole conversation including Mika](https://wordpress.slack.com/archives/C1LBM36LC/p1526324067000418).

We don't have immediate plans to publish to .org (https://github.com/versionpress/versionpress/issues/1291#issuecomment-357602351) but at the same time, 7.0 language level should give us most of what we need and might be more practical. There are subtle issues like the [current PHPUnit 7 only supporting 7.1 and newer](https://phpunit.de/) but we should be able to stay with v6 just fine.

Let's see, this will be for a separate issue / PR.

**Comment by borekb** [comment]
I spent some time yesterday trying to fix Selenium tests and it will be more involved it seemed, for a separate PR. Some notes:

- The key dependencies (Selenium / Webdriver / Firefox images) updated significantly during the past year and were not work for me out of the box.
- Firefox was not connecting at all, switching to Chrome `selenium/standalone-chrome` was better but was also failing on things that should just work, for example, maximizing window (maybe [this chromedriver bug](https://bugs.chromium.org/p/chromedriver/issues/detail?id=1901)?).
- Some code around Selenium tests, sometimes but not always dictated by the `giorgiosironi/phpunit-selenium` library we use, looked really quite weird to me, e.g., [here](https://github.com/versionpress/versionpress/blob/d322b8bdf1be5d4d01eaeacbe09877aaa4e3d424/plugins/versionpress/tests/End2End/Utils/SeleniumWorker.php#L17-L57) or [here](https://github.com/versionpress/versionpress/blob/d322b8bdf1be5d4d01eaeacbe09877aaa4e3d424/plugins/versionpress/tests/End2End/Utils/SeleniumWorker.php#L89-L103). The [actual test code](https://github.com/versionpress/versionpress/blob/d322b8bdf1be5d4d01eaeacbe09877aaa4e3d424/plugins/versionpress/tests/End2End/Options/OptionsTestSeleniumWorker.php#L14-L18) looked fine to me.
- I'm not sure that `giorgiosironi/phpunit-selenium` is still the way to go. `facebook/webdriver` seems to be [more popular](https://packagist.org/?query=selenium) or maybe we could look at frameworks like [Codeception](https://codeception.com/).

**Comment by borekb** [comment]
d6cdb26 is an important change to the structure of docker-compose testing stack – separate services have been defined for tests and I've also fixed WpAutomation. At this point, **all End2End tests with the WP-CLI worker pass** for me, which if for the first time in this PR.

Also, when working on this, I've encountered two PHPUnit test reporters that could be useful (printing errors immediately instead of waiting for the test suite to finish):

- [`scriptfusion/phpunit-immediate-exception-printer`](https://packagist.org/packages/scriptfusion/phpunit-immediate-exception-printer)
- [`rpkamp/fancy-testdox-printer`](https://packagist.org/packages/rpkamp/fancy-testdox-printer) – built in in PHPUnit 7+.

**Comment by borekb** [comment]
Status update:

db6842c adds support for command-line started debugging. That will be quite important for me as I cannot get PhpStorm to work correctly with Docker Compose ([IDEA-192416](https://youtrack.jetbrains.com/issue/IDEA-192416)).

All tests are passing for me _when_ I launch them in smaller batches. The full `tests:full` suite is failing on too many database connections, we're probably not clearing things up properly somewhere. `DBAsserter` and it's [`staticInitialization()`](https://github.com/versionpress/versionpress/blob/dc531ed1771a9c83170f435fd3956588399678e7/plugins/versionpress/tests/Utils/DBAsserter.php#L52) method is my guess. This query:

```
show status where `variable_name` = 'Max_used_connections';
```

returned like 150 concurrent connections.

**Comment by borekb** [comment]
2a477de makes Workflow tests run again.

Merge driver test is still failing because it seems that the merge driver doesn't select the newer datetime – might be an actual bug in it. This is an example output:

```
1) VersionPress\Tests\Workflow\CloneMergeTest::dateModifiedMergesAutomatically
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'2018-05-23 10:06:38
+'2018-05-23 10:05:51
 '
```

(The newer date is expected but `wp vp pull` and the merge driver chooses the older one.)

**Comment by borekb** [comment]
🎉 WP-CLI tests now fully now! The only failures are merge driver tests as described in https://github.com/versionpress/versionpress/pull/1329#issuecomment-391311790 which will be for a separate PR.

image

Remaining TODO items in this PR:

- [ ] Try to speed up tests by avoiding volumes (they used to be very slow on Windows, see e.g. https://github.com/docker/for-win/issues/188). Or things like http://docker-sync.io/ or [LCOW](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/linux-containers).
- [ ] Final verification and cleanup of `Dev-Setup.md`.
    - [ ] @pavelevap to possibly contribute tips for Docker Toolbox.
- [ ] Maybe write some output for long-running operations like initial site setup; the console is "dark" for quite a while.

**Comment by borekb** [comment]
I've updated the PR description and consider this PR mostly done. @JanVoracek verified that the tests run exactly the same on Mac which is good.

If someone wants to look at Dev-Setup.md at this point and possibly try the dev / test setup, that would be great. I think we'll be able to merge next week.

**Comment by borekb** [comment]
Thanks, @pavelevap, I think we can merge this.

**Pull request merged**